### PR TITLE
[PR-C1] feat(core): implement semantic identities, kinds, schemas, and shapes

### PR DESCRIPTION
**Pull request by cmontella**
## Stack

C0 base branch: `test/value-semantic-migration-contracts`  
C0 exact base SHA: `f89d369ddd356d31abce6d79de53fb0cc811a83d`  
C1 branch: `feat/core-semantic-foundations`  
C1 exact head SHA: `b52d369e93a1d26b27bef476329bc54cc9966a64`

## Purpose

Implement final semantic identities, dimensions, `KindExpr`, `KindScheme`, `Schema`, `ShapeInstance`, `SchemaKey`, `NominalKey`, `SchemaTable`, and explicit legacy adapters.

## Production routing

No compiler or runtime route changes. The current mutable `Value` remains authoritative.

## Validated construction boundary

- Open semantic syntax retains `Serialize + Deserialize`.
- Finalized `Schema`, `ShapeInstance`, `KindScheme`, `CanonicalNominalPath`, and `DimensionParameter` are `Serialize`-only and can be created only through validated APIs; derived, manual, qualified, renamed, and target-aliased `Deserialize` implementations are rejected, including for `SchemaHandle`.
- Semantic builders return structured errors for compile-time parameters and exhausted dimension/schema handles.
- Dimension expressions and parameter bounds normalize before parameter reachability is computed, so algebraically eliminated references do not survive in canonical identities.
- `SchemaHandle` is non-Serde, carries the inserted schema identity, and resolves through checked bounds/identity validation rather than unchecked indexing.
- `KindExpr` rejects duplicate record fields and table columns recursively through schemes, canonical closed-kind encoding, and legacy adapters.
- Shape instantiation checks the total product of separate matrix dimensions recursively.
- Legacy context output validates every declaration and returned dimension expression, including unused declarations.
- C1 enforcement resolves transitive aliases/wrappers, rejects standard interior-mutability dependencies, and structurally rejects wildcard, binding, attributed, parenthesized, and irrefutable or-pattern catch-all adapter arms.

## Identity model

- `ProgramRevision`: durable content-addressed program revision.
- `SchemaKey`: SHA-256 identity of canonical finalized schema bytes.
- `NominalKey`: SHA-256 identity of canonical atom/enum kind and path.
- `KindId`: deterministic kind-table identity.
- `KindParameterId`: dense quantified kind-parameter identity.
- `DimensionParameterId`: dense dimension-parameter identity.
- `SchemaId`: deterministic schema-table-local identity.
- `ReactiveInstanceId`: instance index plus checked generation.
- `CellSlotId`: stable cell slot within a reactive instance.
- `CellId`: reactive instance plus cell slot.
- `SlotIndex`: resident storage slot index.
- `InstanceEpoch`: checked resident publication epoch.
- `PlanGeneration` / `LayoutGeneration`: checked plan and layout generations.

## Canonical conformance

Positive schema vectors reproduced: 11 / 11  
Dimension vectors reproduced: 7 / 7  
Invalid schema vectors reproduced: 8 / 8  
SchemaKey vectors reproduced: 11 / 11  
NominalKey vectors reproduced: 1 / 1

## Legacy adapters

Kind variants covered: 17 / 17  
ValueKind variants covered: 32 / 32  
Wildcard, binding, or attributed catch-all arms: 0  
Ambiguous dimensions guessed: 0

The nonblocking adapter-error taxonomy cleanup is tracked separately in #737.

## Migration progress

C1-gated targets implemented: 37 / 37  
C2/C3/C4/D targets implemented: 0  
Legacy types removed: 0

## Resident compatibility

CellSlotId size: 4 bytes (`u32`)  
SlotIndex size: 4 bytes (`u32`)  
InstanceEpoch size: 8 bytes (`u64`)  
Steady-state allocations: 0 allocations / 0 bytes per turn  
Publication stores: 1 per accepted turn

## Gate B

Implementation commit measured: `501a40d9b1e634c62d5c8565aa2a14ff9aa13237`  
Evidence/report commit: `b52d369e93a1d26b27bef476329bc54cc9966a64`  
Report `evidence_commit`: `501a40d9b1e634c62d5c8565aa2a14ff9aa13237`  
Report SHA-256: `12cc4c4940752fd864702ac3b0a80a2d60fde064442837b5f3beb0a36b1d0b87`  
Raw-epoch ratio: `1.2113950645` (limit `1.25`)  
Legacy-gap closure: `0.9937288494` (minimum `0.80`)  
History ratios: `1k/0 = 1.0031794571`, `100k/0 = 1.0063871011`, high/low epoch `= 0.9979227382` (limits `1.05`)  
Primary denominator: `8249.2476654 ns/turn` (positive)  
Decision: `Pass`

All hard gates passed: correctness, executor tax, legacy-gap closure, zero allocation, no full clone, constant publication, history independence, tail stability, infallible post-publication append, and raw-epoch ratio.

## Commits

1. `d3a6520a04d4cc375c475ed1fea335abf62e3621` — `feat(core): add final semantic identities`
2. `d7313708973f2cc5dec92f2693452e91959cce10` — `feat(core): add dimensions and kind schemes`
3. `5ecb35161ac1bc8d9e81557a80e98cd8ef6892f6` — `feat(core): add canonical schemas and schema tables`
4. `6fefebafcbb071f008d9af45fd2d0444bd72813b` — `refactor(core): add explicit legacy kind adapters`
5. `501a40d9b1e634c62d5c8565aa2a14ff9aa13237` — `test(architecture): prove C1 semantic and resident contracts`
6. `b52d369e93a1d26b27bef476329bc54cc9966a64` — `bench(architecture): refresh Gate B evidence after C1`

## Validation

- [x] `cargo fmt --all -- --check`
- [x] `cargo check -p mech-core --no-default-features`
- [x] `cargo check -p mech-core --no-default-features --features no_std`
- [x] `cargo test -p mech-core --all-features` — 285 unit tests plus all C1 integration contracts passed
- [x] focused resident engine tests — 5 passed
- [x] resident EKF contract tests — 4 passed
- [x] `cargo test -p mech-runtime --lib --no-default-features --features source_default` — 660 passed
- [x] current inventory and frozen legacy-baseline checks
- [x] `python3 scripts/check-value-system-contract.py` — passed in 307.5s with single-pass production-source analysis
- [x] `python3 scripts/check-value-execution-boundary.py`
- [x] exact-commit `python3 scripts/check-gate-b-contract.py --report ... --expected-commit ...`, including independent ancestry verification against the frozen B2 floor
- [x] generator/runner/Gate B/value-system checker suites — 253 passed
- [x] independent canonical-encoding vector reproduction
- [x] resident EKF benchmark target build
- [x] standard distribution contract, packaging, native-host catalog, and static-profile checks
- [x] controlled B2 benchmark on MacBook Air (Mac15,13), Apple M3, 8 cores, 16 GB
- [x] Static contracts retains every check with temporary 15-minute stabilization headroom
- [x] exactly six commits over the exact C0 base
- [x] clean worktree and `git diff --check`

### CODE REVIEW

**Comment by cmontella** [comment]
### C1 blocker — finalized invariant-bearing types must not derive unchecked `Deserialize`

The current Serde derives can construct finalized types without running their validation boundary. In particular, derived `Deserialize` can bypass `SchemaDraft::finalize()`, `Schema::instantiate_shape()`, `KindScheme::new()`, and `CanonicalNominalPath::new()`.

Required change:

1. Keep `Serialize + Deserialize` on open/draft syntax types such as `SchemaDraft`, `SchemaBody`, `DimensionExpr`, `DimensionParameterDeclaration`, `KindExpr`, `KindParameter`, `InputKindScheme`, and `KindConstraint`.
2. Remove derived `Deserialize` from these finalized types:
   - `Schema`
   - `ShapeInstance`
   - `KindScheme`
   - `CanonicalNominalPath`
   - `DimensionParameter`
3. `ShapeInstance` must remain non-deserializable without a `Schema`; validation requires schema parameter count and bounds.
4. Do not add unchecked private-field wire constructors. A later gate may add explicit wire DTOs whose conversion invokes the validated constructors.
5. `Serialize` may remain on finalized types.

Add tests/contract enforcement proving:

- invalid serialized data cannot directly construct a finalized `Schema`, `ShapeInstance`, `KindScheme`, `CanonicalNominalPath`, or `DimensionParameter`;
- drafts/open expressions still support Serde where enabled;
- every finalized instance used by canonical encoding was produced through its validation API.

A small trybuild compile-fail test requiring `DeserializeOwned` for the finalized types is acceptable and avoids relying only on source grep.

**Comment by cmontella** [comment]
### C1 blocker — semantic construction APIs must return structured errors, not panic

`DimensionEnvironmentBuilder::declare()` currently uses `assert!` for compile-time parameters and `expect` for ID exhaustion. `SchemaTableBuilder::insert()` also uses `expect` for handle exhaustion. These APIs consume compiler/context-controlled semantic input and must not terminate the process.

Required API changes:

```rust
pub fn declare(
    &mut self,
    origin: DimensionParameterOrigin,
    lifetime: DimensionLifetime,
    lower_bound: DimensionExpr,
    upper_bound: Option<DimensionExpr>,
) -> Result<DimensionParameterId, SemanticModelError>;

pub fn insert(
    &mut self,
    schema: Schema,
) -> Result<SchemaHandle, SemanticModelError>;
```

Use these exact errors:

- `DimensionLifetime::CompileTime` passed to `declare()` -> `SemanticModelError::CompileTimeDimensionParameterV1`.
- dimension parameter ordinal exhaustion -> `SemanticModelError::IdentityExhausted { identity: SemanticIdentityKind::DimensionParameterId }`.
- schema insertion-handle exhaustion -> `SemanticModelError::IdentityExhausted { identity: SemanticIdentityKind::SchemaHandle }`.

Add `DimensionParameterId` and `SchemaHandle` to `SemanticIdentityKind`.

Update all callers, including `LegacySemanticContext::resolve_unspecified_extent`, fake contexts, legacy adapters, schema-table tests, and builders, to propagate `?`. Do not translate these failures into `LegacyExtentResolutionKindMismatch`.

For practical exhaustion tests, add private/test-only limit helpers rather than allocating billions of entries. Tests must prove:

- compile-time parameter declaration returns the structured error and does not panic;
- dimension parameter handle exhaustion returns the structured error;
- schema handle exhaustion returns the structured error;
- existing schema unique-ID exhaustion remains `SchemaIdExhausted` during `finish()`;
- no production semantic builder uses `assert!`, `panic!`, or `expect` for input-derived cardinalities or identities.

**Comment by cmontella** [comment]
### C1 blocker — validate the checked product of matrix dimensions

`Schema::instantiate_shape()` currently evaluates every matrix dimension independently but does not validate the matrix’s total logical element count. A schema with dimensions `[u64::MAX, 2]` therefore passes even though its cardinality overflows.

Required change in `schema/shape.rs`:

```rust
SchemaBody::Matrix {
    element,
    dimensions,
} => {
    evaluate_body_extents(element, values)?;

    let mut element_count = 1_u64;
    for dimension in dimensions {
        let extent = evaluate_dimension(dimension, values)?;
        element_count = element_count
            .checked_mul(extent)
            .ok_or(SemanticModelError::DimensionOverflowV1)?;
    }
}
```

The count need not be stored in `ShapeInstance`; this is validation only. Rank-zero matrices retain the mathematical empty-product value `1` unless the frozen schema contract separately prohibits rank zero.

Apply this recursively, so matrices nested in tuples, records, options, enum payloads, tables, sets, or maps receive the same validation.

Add tests for:

- two individually valid constant dimensions whose product overflows;
- parameterized dimensions whose resolved product overflows;
- a large but non-overflowing product that succeeds;
- a nested matrix payload that triggers the same error.

The existing test for overflow inside one table row-count expression is not sufficient; the new test must exercise multiplication across separate matrix dimension entries.

**Comment by cmontella** [comment]
### C1 blocker — reject duplicate record fields and table columns in `KindExpr`

`Schema` rejects duplicate names, but validated/canonical kind expressions currently accept and encode duplicate `KindField` names. That allows a `KindScheme` or closed kind that can never resolve to a valid schema.

Add:

```rust
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum KindNameCategory {
    RecordField,
    TableColumn,
}

SemanticModelError::DuplicateKindName {
    category: KindNameCategory,
    name: String,
}
```

Implement one recursive validator in `kind_expr.rs`, for example:

```rust
pub(crate) fn validate_kind_structure(
    kind: &KindExpr,
) -> Result<(), SemanticModelError>;
```

Required behavior:

- `KindExpr::Record`: field names must be unique in declaration order; recurse into every field kind.
- `KindExpr::Table`: column names must be unique in declaration order; recurse into every column kind and validate the row dimension through the existing dimension path.
- recurse through matrix element, option, tuple, set, map key/value, reference, and `TypeOf`.
- do not sort record/table fields; order remains semantic.

Invoke this validator from all construction/finalization boundaries that claim a valid kind structure:

1. `KindScheme::new()` for kind-parameter bounds, inputs, outputs, and kind constraints;
2. `canonical_closed_kind_bytes()` before canonicalization/encoding;
3. `kind_expr_from_legacy()` before returning `LegacyKindResolution`.

Add tests proving duplicate names fail through all three routes, including a duplicate nested inside another aggregate. Distinct names with reversed order must remain valid and produce different canonical bytes where order is semantic.

**Comment by cmontella** [comment]
### C1 correctness note — validate `LegacySemanticContext` output before returning it

`schema_from_legacy_value_kind()` eventually passes through `SchemaDraft::finalize()`, but `kind_expr_from_legacy()` currently returns the context-produced `KindExpr` and declarations directly. A faulty context can return `DimensionExpr::Parameter(999)` without declaring that parameter, or add invalid/cyclic bounds, and the adapter will return an inconsistent `LegacyKindResolution`.

Required change:

Add an internal validation step before `kind_expr_from_legacy()` returns. It must:

1. run `validate_kind_structure(&kind)`;
2. require declaration IDs to be dense and unique;
3. validate **all** declared parameter bounds, including unused declarations, for compile-time-parameter misuse, unknown references, forward references, cycles, holes, and arithmetic overflow;
4. traverse every dimension expression in the returned `KindExpr` and call `normalize_dimension(expr, declarations.len())` to prove every reference is in range and no hole remains where the adapter claims a resolved extent;
5. return the existing structured `SemanticModelError` without guessing or rewriting a missing context declaration.

A concrete implementation may call `canonicalize_dimension_environment()` once with all declared IDs to validate the complete declaration set, then separately validate dimensions reachable from the kind. It does not need to replace the returned draft declarations with finalized `DimensionParameter` values.

Add malicious-context tests for:

- returning `Parameter(999)` without using the supplied builder;
- returning a reference to a parameter not yet declared;
- creating cyclic parameter bounds;
- returning `DimensionExpr::Hole` as a supposedly resolved extent;
- a valid context-created parameter succeeding.

This keeps the context extensible while making the adapter boundary trustworthy.

**Comment by cmontella** [comment]
### Non-blocking follow-up — keep transitional legacy-adapter errors out of the permanent semantic error API

This is not a C1 merge blocker and should not delay the four correctness fixes above. The current `SemanticModelError` includes adapter-only concepts such as `LegacyExtentSite` and `LegacyValueKindTag`, which should not become permanent semantic-model vocabulary.

Preferred follow-up design:

```rust
pub enum LegacyAdapterError {
    Semantic(SemanticModelError),
    NamedKindUnresolved { legacy_id: u64 },
    NominalUnresolved { ... },
    ExtentUnresolved { site: LegacyExtentSite },
    ExtentOutOfRange { value: u128 },
    ExtentResolutionKindMismatch,
    NonInstantiableValueKind { kind: LegacyValueKindTag },
}
```

Then:

- move all `Legacy*` variants out of `SemanticModelError`;
- make `kind_expr_from_legacy`, `schema_from_legacy_value_kind`, and `LegacySemanticContext` return `LegacyAdapterError`;
- implement `From<SemanticModelError> for LegacyAdapterError`;
- implement `MechErrorKind` and `From<LegacyAdapterError> for MechError` at the boundary;
- keep canonical schema/dimension/identity code dependent only on `SemanticModelError`.

Handle this in C1 only if it remains a small, isolated refactor while fixing the adapter validation note. Otherwise record it as C1.1/C2 cleanup; do not expand this PR or rerun Gate B solely for this error-taxonomy split.

**Comment by cmontella** [comment]
### Integration instructions after addressing the C1 notes

Keep the stack at exactly six commits; do not append a seventh review-fix commit.

Amend/autosquash fixes into their owning commits:

- finalized Serde invariants and identity/error changes -> commits 1–3 as appropriate;
- checked dimension builder and kind validation -> commit 2;
- matrix extent validation and schema-table insertion API -> commit 3;
- legacy context-output validation -> commit 4;
- architecture/static-contract enforcement and regression fixtures -> commit 5.

Because commits 1–5 will change, the existing Gate B report is no longer attributable to the final implementation. Delete/discard the current commit-6 evidence changes, finish and freeze the new commit 5, then rerun the complete controlled `B2-resident-turn` benchmark against that exact implementation SHA. Recreate commit 6 with only the refreshed report and permitted evidence metadata.

Do not weaken the performance thresholds. The final report must still prove:

```text
raw_epoch_ratio <= 1.25
legacy_gap_closure >= 0.80
zero steady-state allocation
one publication store
history and high-epoch ratios <= 1.05
infallible post-publication append
```

Resolve these notes only after focused tests pass and the refreshed exact-head CI run is green.

**Comment by cmontella** [comment]
Exact-head resolution summary for ae45471f2d58489ccb460ba9882db79c1bb468e0:

- Removed Deserialize from finalized Schema, CanonicalNominalPath, ShapeInstance, and KindScheme boundaries while retaining it for open/draft syntax.
- Expanded legacy-adapter catch-all enforcement to reject wildcard, binding, and binding-at-wildcard arms.
- Made the C1 semantic-dependency guard resolve transitive aliases, wrapper types, and renamed imports using repository parser helpers.
- Also completed the structured-builder, checked matrix-product, recursive duplicate-name, and full legacy-context validation findings.
- Deferred the nonblocking adapter error-taxonomy expansion to #737 because it is outside C1.

All exact-head CI checks are green, including Static contracts. The six superseded inline findings are resolved.

**Comment by cmontella** [comment]
Round 1 resolution summary for exact head 324bbf3a075ad7b11c42a2254ac05f439f3e6d84:

- Dimension expressions in schema bodies, closed kinds, and parameter bounds now normalize before reachability, with regressions proving algebraically eliminated references do not alter canonical keys or require shape values.
- Legacy-adapter enforcement now skips arbitrary balanced outer attributes before detecting wildcard and binding catch-all arms; bare, attributed, cfg_attr, multiple-attribute, ref/mut, and binding-at-wildcard regressions pass.
- SchemaHandle no longer derives Serde, carries the inserted SchemaKey, and resolve returns a structured checked result that rejects out-of-range and wrong-schema foreign handles without indexing panics.
- The descendant Gate B evidence validator now accepts a C1 refresh only when an exact expected commit is supplied; historical branch validation remains frozen.

All local suites and exact-head CI are green. Gate B decision is Pass. The three round-1 threads are resolved.

File: src/core/src/schema/mod.rs [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+use std::{boxed::Box, string::String};
+
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct SchemaDraft {
+    pub dimension_parameters: Box<[DimensionParameterDeclaration]>,
+    pub body: SchemaBody,
+}
+
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
```
[the remark this answers is not in the document]
Comment (cmontella) [reply src/core/src/schema/mod.rs]: Addressed in commit 3 (`5ecb35161ac1bc8d9e81557a80e98cd8ef6892f6`) and enforced in commit 5. Finalized `Schema` is Serialize-only; deserialization remains on `SchemaDraft`, whose `finalize()` validates the invariant boundary. Serde contract tests and exact-head CI pass.

File: src/core/src/schema/shape.rs [hunk 2]
```diff
@@ -0,0 +1,187 @@
+use super::{Schema, SchemaBody};
+use crate::{DimensionExpr, DimensionOperator, DimensionParameterId, SemanticModelError};
+
+#[cfg(feature = "no_std")]
+use alloc::{boxed::Box, vec::Vec};
+#[cfg(not(feature = "no_std"))]
+use std::{boxed::Box, vec::Vec};
+
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
```
[the remark this answers is not in the document]
Comment (cmontella) [reply src/core/src/schema/shape.rs]: Addressed in commit 3 (`5ecb35161ac1bc8d9e81557a80e98cd8ef6892f6`) and enforced in commit 5. `ShapeInstance` is Serialize-only and can only be produced by `Schema::instantiate_shape`, which validates parameter count, bounds, and recursive extent products. Exact-head CI passes.

File: src/core/src/nominal.rs [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+use crate::{NominalKey, NominalPathError, SemanticModelError};
+use sha2::{Digest, Sha256};
+
+#[cfg(feature = "no_std")]
+use alloc::{boxed::Box, string::String, vec::Vec};
+#[cfg(not(feature = "no_std"))]
+use std::{boxed::Box, string::String, vec::Vec};
+
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
```
[the remark this answers is not in the document]
Comment (cmontella) [reply src/core/src/nominal.rs]: Addressed in commit 1 (`d3a6520a04d4cc375c475ed1fea335abf62e3621`) and enforced in commit 5. `CanonicalNominalPath` is Serialize-only, so construction must pass through `CanonicalNominalPath::new`. Serde boundary regressions and exact-head CI pass.

File: scripts/check-value-system-contract.py [hunk 4]
```diff
@@ -1419,6 +1465,70 @@ def future_boundary_failures(root: Path) -> list[Failure]:
...
+                "ValueStateJournal",
+                "ReactiveTurnJournal",
+                "RuntimeExecutionTransaction",
+                "StateArena",
+                "nalgebra",
+                "DMatrix",
+            }
+            forbidden = next(
+                (token for token in tokens if token.value in forbidden_identifiers),
+                None,
```
[the remark this answers is not in the document]
Comment (cmontella) [reply scripts/check-value-system-contract.py]: Addressed in commit 5 (`571622afe8b44092be6eaf58f623384055d120cb`). C1 semantic dependency enforcement now resolves transitive type aliases, wrapper fields/payloads, and renamed imports using the repository Rust-token helpers. The alias/wrapper regressions and exact-head CI pass.

File: scripts/check-value-system-contract.py [hunk 5]
```diff
@@ -1419,6 +1465,70 @@ def future_boundary_failures(root: Path) -> list[Failure]:
...
+                        "forbidden production dependency",
+                        "src/core/src/legacy_adapter/",
+                        forbidden.line
+                        if forbidden is not None
+                        else source.count("\n", 0, physical.start()) + 1,
+                        forbidden.column if forbidden is not None else None,
+                    )
+                )
+        if module_file_or_descendant(root, resolved, "legacy_adapter"):
+            wildcard = re.search(r"(?:^|[,{}])\s*_\s*=>", searchable)
```
[the remark this answers is not in the document]
Comment (cmontella) [reply scripts/check-value-system-contract.py]: Addressed in commit 5 (`571622afe8b44092be6eaf58f623384055d120cb`). The legacy-adapter exhaustiveness guard rejects wildcard, binding, `ref`/`mut`, binding-at-wildcard, and attributed fallbacks. Regression fixtures cover each form and the 221-test script suite plus exact-head CI pass.

File: scripts/check-value-system-contract.py [hunk 6]
```diff
@@ -1410,15 +1456,365 @@ def module_file_or_descendant(root: Path, resolved: Path, name: str) -> bool:
...
+                        if direct_forbidden is not None
+                        else transitive_token.column
+                        if transitive_token is not None
+                        else None,
+                    )
+                )
+        if module_file_or_descendant(root, resolved, "legacy_adapter"):
+            catch_all = re.search(
+                r"(?:^|[,{}])\s*(?:(?:ref|mut)\s+)*(?:_|r#[a-z_][A-Za-z0-9_]*|[a-z_][A-Za-z0-9_]*)\s*(?:@\s*_\s*)?=>",
+                searchable,
```
[the remark this answers is not in the document]
Comment (cmontella) [reply scripts/check-value-system-contract.py]: Addressed in commit 5 (`571622afe8b44092be6eaf58f623384055d120cb`). The guard now skips repeated balanced outer attributes before recognizing a fallback. Tests cover `allow`, `cfg_attr`, multiple attributes, `mut`, `ref`, and `other @ _`; the full script suite and exact-head CI pass.

File: src/core/src/schema/table.rs [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+        assert!(matches!(
+            builder.finish_with(u32::MAX as usize, |_| forced),
+            Err(SemanticModelError::SchemaKeyCollision { key }) if key == forced
+        ));
+    }
+}
+
+impl SchemaTableBuild {
+    pub fn resolve(&self, handle: SchemaHandle) -> SchemaId {
+        self.remap[handle.0 as usize]
```
[the remark this answers is not in the document]
Comment (cmontella) [reply src/core/src/schema/table.rs]: Addressed in commit 3 (`5ecb35161ac1bc8d9e81557a80e98cd8ef6892f6`) with durability enforcement in commit 5. `SchemaHandle` is non-Serde, carries the inserted `SchemaKey`, and `resolve` returns a checked `Result` that rejects out-of-range or wrong-schema foreign handles without indexing. Unit and integration regressions plus exact-head CI pass.

File: src/core/src/kind_scheme.rs [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+    DimensionExpr, DimensionParameterDeclaration, DimensionParameterId, KindExpr, KindParameterId,
+    SemanticModelError,
+};
+
+#[cfg(feature = "no_std")]
+use alloc::{boxed::Box, collections::BTreeSet, vec::Vec};
+#[cfg(not(feature = "no_std"))]
+use std::{boxed::Box, collections::BTreeSet, vec::Vec};
+
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
```
[the remark this answers is not in the document]
Comment (cmontella) [reply src/core/src/kind_scheme.rs]: Addressed in commit 2 (`d7313708973f2cc5dec92f2693452e91959cce10`) and enforced in commit 5. `KindScheme` is Serialize-only and construction remains behind `KindScheme::new`, including hole, reference, density, and bound validation. Exact-head CI passes.

File: src/core/src/schema/validation.rs [hunk 9]
```diff
@@ -0,0 +1,387 @@
...
+
+#[cfg(feature = "no_std")]
+use alloc::{boxed::Box, collections::BTreeSet, string::String, vec::Vec};
+#[cfg(not(feature = "no_std"))]
+use std::{boxed::Box, collections::BTreeSet, string::String, vec::Vec};
+
+pub(super) fn finalize_schema(draft: SchemaDraft) -> Result<Schema, SemanticModelError> {
+    validate_names_and_keyability(&draft.body)?;
+    let mut references = Vec::new();
+    collect_body_dimension_references(&draft.body, &mut references);
```
[the remark this answers is not in the document]
Comment (cmontella) [reply src/core/src/schema/validation.rs]: Addressed across commits 2 and 3 (`d7313708973f2cc5dec92f2693452e91959cce10`, `5ecb35161ac1bc8d9e81557a80e98cd8ef6892f6`). Schema bodies, closed kinds, and parameter bounds normalize before reachability. Regressions prove `0 * p` removes `p`, preserves canonical-key equality with constant zero, and does not require a shape value. Gate B was regenerated after the rewrite and exact-head CI passes.

**Comment by cmontella** [comment]
Top-level review-note status (these are conversation comments, so GitHub has no thread Resolve control):

- ✅ [Finalized Serde boundary](https://github.com/mech-lang/mech/pull/735#issuecomment-5229162061) — implemented across commits 1–3 and enforced in commit 5.
- ✅ [Structured semantic builder errors](https://github.com/mech-lang/mech/pull/735#issuecomment-5229162669) — implemented with exhaustion and compile-time regressions.
- ✅ [Checked matrix dimension product](https://github.com/mech-lang/mech/pull/735#issuecomment-5229163193) — implemented recursively with constant, parameterized, valid-large, and nested regressions.
- ✅ [Duplicate KindExpr names](https://github.com/mech-lang/mech/pull/735#issuecomment-5229163697) — implemented at scheme, closed-kind, and legacy-adapter boundaries.
- ✅ [LegacySemanticContext output validation](https://github.com/mech-lang/mech/pull/735#issuecomment-5229164849) — implemented for dense/unique declarations, all bounds, references, cycles, holes, overflow, and returned dimensions.
- ↗️ [Non-blocking legacy-adapter error taxonomy](https://github.com/mech-lang/mech/pull/735#issuecomment-5229165641) — explicitly not a C1 blocker; deferred to #737 as instructed.
- ✅ [Six-commit integration and refreshed evidence](https://github.com/mech-lang/mech/pull/735#issuecomment-5229166238) — exactly six commits, report regenerated from commit 5, all local suites and exact-head CI green.

Inline review status: 9 of 9 threads have individual fix replies and are resolved; 0 unresolved and 0 unanswered.

File: scripts/check-gate-b-contract.py [hunk 10]
```diff
@@ -671,13 +673,19 @@ def report_contract_errors(
...
-        if report["git_branch"] != expected_branch:
+        if (
+            report["phase"] == "B1-resident-kernel"
+            and report["git_branch"] != "feat/engine-resident-ekf-substrate"
+        ):
             errors.append("B1 evidence is attributed to the wrong branch")
+        if (
+            report["phase"] == "B2-resident-turn"
+            and report["git_branch"] != "perf/runtime-resident-ekf-efficacy"
+            and expected_commit is None
```
[the remark this answers is not in the document]
Comment (cmontella) [reply scripts/check-gate-b-contract.py]: Accepted. I am adding validator-side `git merge-base --is-ancestor` verification against `B2_EVIDENCE_FLOOR` for every non-frozen B2 refresh, independent of the runner and in addition to exact SHA equality. Tests will cover valid descendants, unrelated histories, missing commits, and the frozen historical branch.

File: scripts/check-value-system-contract.py [hunk 11]
```diff
@@ -1410,15 +1456,417 @@ def module_file_or_descendant(root: Path, resolved: Path, name: str) -> bool:
...
+    "Ref",
+    "ValRef",
+    "MutableReference",
+    "ReactiveCellId",
+    "ValueStateJournal",
+    "ReactiveTurnJournal",
+    "RuntimeExecutionTransaction",
+    "StateArena",
+    "nalgebra",
+    "DMatrix",
```
[the remark this answers is not in the document]
Comment (cmontella) [reply scripts/check-value-system-contract.py:1541]: Accepted. I am defining and enforcing the standard interior-mutability family (`UnsafeCell`, cell/lock/once/lazy primitives, atomics) as forbidden semantic dependencies, with direct, imported, and transitively wrapped regressions. This remains confined to the C1 finalized-semantic boundary.

File: scripts/check-value-system-contract.py [hunk 12]
```diff
@@ -1410,15 +1456,417 @@ def module_file_or_descendant(root: Path, resolved: Path, name: str) -> bool:
...
+    if start < 0:
+        return None
+    end = searchable.find(end_marker, start + len(start_marker))
+    return searchable[start:] if end < 0 else searchable[start:end]
+
+
+def legacy_adapter_catch_all(source: str) -> int | None:
+    binding = re.compile(
+        r"(?:(?:ref|mut)\s+)*(?:_|r#[a-z_][A-Za-z0-9_]*|[a-z_][A-Za-z0-9_]*)"
+        r"\s*(?:@\s*_\s*)?=>"
```
[the remark this answers is not in the document]
Comment (cmontella) [reply scripts/check-value-system-contract.py]: Accepted. The repeated regex gaps indicate the arm detector needs pattern structure rather than another spelling patch. I am replacing the immediate-binding regex with balanced arm-pattern analysis that strips attributes/outer parentheses, splits top-level or-patterns, and recursively detects irrefutable wildcard or binding alternatives, including attributed and parenthesized forms.

File: scripts/check-value-system-contract.py [hunk 13]
```diff
@@ -1410,15 +1456,417 @@ def module_file_or_descendant(root: Path, resolved: Path, name: str) -> bool:
...
+    corpus = list(production_corpus(root))
+    semantic_forbidden_names = transitive_semantic_forbidden_names(corpus)
+    for path, source in corpus:
         resolved = path.resolve()
         relative = path.relative_to(root).as_posix()
         searchable = GENERATOR.mask_non_code(source)
         tokens = GENERATOR.rust_tokens(source, searchable)
+        for type_name in FINALIZED_SEMANTIC_SERDE_TYPES.get(relative, set()):
+            attribute = semantic_serde_attribute(source, type_name)
+            if attribute is not None and attribute != '#[cfg_attr(feature = "serde", derive(Serialize))]':
```
[the remark this answers is not in the document]
Comment (cmontella) [reply scripts/check-value-system-contract.py:1981]: Accepted. I am extending the Serde boundary from declaration-derive checks to implementation checks across the production corpus, including qualified `serde::de::Deserialize`, renamed imports, and target aliases for every finalized type and `SchemaHandle`. Regression fixtures will prove each bypass is rejected.

File: scripts/check-value-system-contract.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (cmontella) [reply scripts/check-value-system-contract.py]: Completed in commit 5 (`501a40d9b1e634c62d5c8565aa2a14ff9aa13237`). The spelling regex was replaced by balanced match-arm pattern analysis: it skips outer attributes, strips balanced parentheses, handles top-level or-patterns and binding-at subpatterns recursively, and distinguishes guarded arms. Regressions cover irrefutable alternatives in either order, leading/parenthesized forms, attributed forms, and explicit/guarded non-catch-alls. All 253 script tests and exact-head CI pass.

File: scripts/check-gate-b-contract.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (cmontella) [reply scripts/check-gate-b-contract.py]: Completed in commit 5 (`501a40d9b1e634c62d5c8565aa2a14ff9aa13237`). Non-frozen B2 refresh validation now independently runs `git merge-base --is-ancestor` against `B2_EVIDENCE_FLOOR`, in addition to exact report-SHA equality. Regressions cover a valid descendant, unrelated history, a missing commit, and the frozen historical branch. The final exact report passes, and exact-head CI run 31297415460 is fully green.

File: scripts/check-value-system-contract.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (cmontella) [reply scripts/check-value-system-contract.py:1541]: Completed in commit 5 (`501a40d9b1e634c62d5c8565aa2a14ff9aa13237`). The semantic boundary now seeds the standard interior-mutability family, including unsafe/cell/lock/once/lazy primitives and atomics, then resolves renamed imports and wrappers transitively. Direct, imported, and wrapped regression fixtures pass. Production sources are masked/tokenized once per audit without weakening the closure. Exact-head CI run 31297415460 is fully green.

File: scripts/check-value-system-contract.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (cmontella) [reply scripts/check-value-system-contract.py:1981]: Completed in commit 5 (`501a40d9b1e634c62d5c8565aa2a14ff9aa13237`). The Serde boundary now scans implementation headers across the production corpus, resolves qualified and renamed `Deserialize` traits plus target aliases, and rejects implementations for every finalized semantic type and `SchemaHandle`; open/draft types remain allowed. Derived and manual-bypass regressions all pass, along with 253 script tests and exact-head CI.
